### PR TITLE
lib/mbedtls: Update to mbedtls-v3.5.1.

### PR DESCRIPTION
**Pull request by Carglglz**
This is a draft for a possible future `mbedTLS 3.x` migration.
Following [3.0-migration-guide](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/3.0-migration-guide.md) I made the necessary changes to update mbedTLS:
```bash
$ git submodule
...
 8c89224991adff88d53cd380f42a2baa36f91454  lib/mbedtls (v3.3.0)
...
```
Unix port builds:
- Linux: OK
- macOS: OK ~(although needs adding `#if...#endif` in `lib/mbedtls/x509.c:137` to avoid `error: unused function 'md_type_to_string'`~

Tests:

```
$ ./run-test.py 
....
837 tests performed (24562 individual testcases)
837 tests passed
....
```
Multi:
```
$ ./run-multitests.py multi_net/*.py
multi_net/ssl_cert_rsa.py on micropython|micropython: pass
multi_net/ssl_context_rsa.py on micropython|micropython: pass
multi_net/ssl_data.py on micropython|micropython: pass
multi_net/tcp_accept_recv.py on micropython|micropython: pass
multi_net/tcp_client_rst.py on micropython|micropython: pass
multi_net/tcp_data.py on micropython|micropython: pass
multi_net/uasyncio_tcp_client_rst.py on micropython|micropython: pass
multi_net/uasyncio_tcp_close_write.py on micropython|micropython: pass
multi_net/uasyncio_tcp_readall.py on micropython|micropython: pass
multi_net/uasyncio_tcp_readexactly.py on micropython|micropython: pass
multi_net/uasyncio_tcp_readinto.py on micropython|micropython: pass
multi_net/uasyncio_tcp_server_client.py on micropython|micropython: pass
multi_net/udp_data.py on micropython|micropython: pass
13 tests performed
13 tests passed
```
For reference esp-idf v.5 will use `mbedTLS 3.x` too [mbed-tls-support-in-esp-idf](https://github.com/espressif/mbedtls/wiki#mbed-tls-support-in-esp-idf).

### TLDR:

Big changes affect configuration options in `mbedtls_config.h`, some functions renamed in `extmod/moduhashlib.c` and 
error codes deprecated/ renamed in `lib/mbedtls_errors/mp_mbedtls_errors.c`. 
Other than that, shouldn't be too difficult to update 👍🏼 .

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for this.  It's a good change but let's wait until after the next release (after v1.20).

**Comment by Carglglz** [comment]
> Thanks for this. It's a good change but let's wait until after the next release (after v1.20).

Yes, it's just something I bumped into while doing #8968 and I figured out it would be nice to have at least some reference when the time comes.

**Comment by Carglglz** [comment]
#### UPDATE:
* I had to change `mbedtls_config.h` to `mbedtls_mpconfig.h` to avoid name conflicts with `mbedtls_config.h` file from `lib/mbedtls` (see [the configuration file](https://mbed-tls.readthedocs.io/en/latest/kb/compiling-and-building/how-do-i-configure-mbedtls/#the-configuration-file)).
* I had to delete the `tests/multi_net/ssl_data.py` test since the option that made the it possible i.e. `# MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_KEY_EXCHANGE`  is deprecated.

**Comment by Carglglz** [comment]
> Thanks for this. It's a good change but let's wait until after the next release (after v1.20).

@dpgeorge I think this is ready to be reviewed, to sum up the changes:

- `mbedtls` source files renamed or deprecated.
- `mbedtls_config.h` renamed to `mbedtls_mpconfig.h` see [the-configuration-file](https://mbed-tls.readthedocs.io/en/latest/kb/compiling-and-building/how-do-i-configure-mbedtls/#the-configuration-file) (there is a `MBEDTLS_CONFIG_FILE` macro but I couldn't make it work, in case you want to test it)
- `MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_KEY_EXCHANGE` is deprecated (so I had to delete this `tests/multi_net/ssl_data.py` test, see https://github.com/micropython/micropython/pull/9506)
- `MBEDTLS_HAVE_TIME` now requires `mbedtls_ms_time` function to be defined but it's only used for `TLSv1.3` (currently not enabled in MicroPython so I added a lazy implementation i.e. `seconds * 1000` that could be improved in the future)
- Tests now use an `EC` key by default and I had to regenerate the `RSA` key due to the old `PKCS` encoding used by `openssl rsa` command, see https://stackoverflow.com/questions/40822328/openssl-rsa-key-pem-and-der-conversion-does-not-match (I've updated `tests/README.md` accordingly)

If there is anything else I can do let me know 👍🏼

**Comment by dpgeorge** [comment]
> `mbedtls_config.h` renamed to `mbedtls_mpconfig.h` see [the-configuration-file](https://mbed-tls.readthedocs.io/en/latest/kb/compiling-and-building/how-do-i-configure-mbedtls/#the-configuration-file) (there is a `MBEDTLS_CONFIG_FILE` macro but I couldn't make it work, in case you want to test it)

Hmm, that's annoying that it's renamed to exactly what we name our config file!

I suggest calling ours `mbedtls_config_port.h`.  That matches the naming style of `mbedtls_config_common.h` and `mbedtls_config_board.h` that are already used in the MicroPython repo.

**Comment by dpgeorge** [comment]
> `MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_KEY_EXCHANGE` is deprecated (so I had to delete this `tests/multi_net/ssl_data.py` test

Does that test really need to be deleted, or is it possible to just update the key/cert data to something else?  The aim of that test is to do a very simple `ssl.wrap_socket()` call and it doesn't really matter what certificates are used.

Well, I guess there is the existing `ssl_cert_rsa.py` and `sslcontext_server_client.py` tests which do very similar, simple SSL data transfer.  So maybe those two are enough and indeed `ssl_data.py` can be removed.

**Comment by Carglglz** [comment]
> Well, I guess there is the existing `ssl_cert_rsa.py` and `sslcontext_server_client.py` tests which do very similar, simple SSL data transfer. So maybe those two are enough and indeed `ssl_data.py` can be removed.

Yes that was my conclusion too. The key/cert there was only 512 bits and there is no clue about how it was generated...

**Comment by Carglglz** [comment]
Also I don't have any device available to test these changes so it would be better if you or anyone could test this on mbedtls ports before merging 👍🏼

**Comment by dpgeorge** [comment]
I have tested this on unix, esp32, rp2 (Pico W) and stm32 (PYBD-SF6) and all SSL-related tests pass.  The tests I ran were from `tests/extmod`, `tests/net_*` and `tests/multi_net`.  I also tested a `requests` get from a HTTPS site.  Everything works!

I have also split this PR up into 3 commits: update mbedtls, update mbedtls_errors, and then the rest of the changes.

**Comment by dpgeorge** [comment]
Now merged!

Thanks @Carglglz for sticking with this for over a year and a half!

**Comment by Carglglz** [comment]
> Thanks @Carglglz for sticking with this for over a year and a half!

Happy to see MicroPython up to date with MbedTLS now! 🎉 

Also I think #11355 can be closed now 👍🏼

**Comment by orangepizza** [comment]
how long it will land on a release branch? it looks like it didn't landed on 1.23.0-preview, so I guess at least 1.24?

**Comment by robert-hh** [comment]
This PR is merged. So it is present in the v1.23 preview versions.

**Comment by orangepizza** [comment]
this PR moved mbedtls_config.h to mbedtls_config_port.h, but that didn't happen in 1.23.0-preview tag and still have havege_c which not valid in mbedtls 3.X
https://github.com/micropython/micropython/blob/v1.23.0-preview/ports/unix/mbedtls/mbedtls_config.h

**Comment by robert-hh** [comment]
I see in the actual preview e.g the file stm32/mbedtls/mbedtls_config_port.h. At which port & version do you look?

**Comment by orangepizza** [comment]
I only see that on master branch but that path still has old config.h in 1.23 preview tag: not but it isn't a branch so maybe it will branch at later point?

**Comment by robert-hh** [comment]
Please show the full path name of the file you are concerned about.

**Comment by orangepizza** [comment]
the link I saw you is already full path? ports/unix/mbedtls/mbedtls_config.h

**Comment by robert-hh** [comment]
The actual state of the repository shows:
https://github.com/micropython/micropython/blob/master/ports/unix/mbedtls/mbedtls_config_port.h
So maybe you have to update your fork  and/or clone of the repository.

**Comment by orangepizza** [comment]
ha x.preview tag is just after x-1 release, well before actual version x branches, never mind
